storage/indexer.go

```diff
@@ -6,20 +6,142 @@ package storage
+	return &Indexer{
+		storageClient: storageClient,
+		options:       options,
+	}
+}
+
+func (i *Indexer) Init(ctx context.Context) error {
+	logger := util.Logger()
+	logger.Debug("Initialize storage indexer")
+
+	err := validateIndexerOptions(i.options)
+	if err != nil {
+		return errors.Wrapf(err, "validation failed")
+	}
+
+	// Populate index file for the first time.
+	err = i.updateIndex(ctx)
+	if err != nil {
+		logger.Error("can't update index file", zap.Error(err))
+	}
 
-func NewIndexer() *Indexer {
-	return new(Indexer)
+	go i.watchIndices(ctx)
+	return nil
 }
 
-func (i *Indexer) Init(context.Context) error {
+func validateIndexerOptions(options IndexerOptions) error {
+	if !strings.HasPrefix(options.PackageStorageBucketInternal, "gs://") {
+		return errors.New("missing or invalid options.PackageStorageBucketInternal")
+	}
+	if !strings.HasPrefix(options.PackageStorageBucketPublic, "gs://") {
+		return errors.New("missing or invalid options.PackageStorageBucketPublic")
+	}
+	if options.WatchInterval == 0 {
+		return errors.New("options.WatchInterval must be greater than 0")
+	}
 	return nil
 }
 
-func (i *Indexer) Get(context.Context, *packages.GetOptions) (packages.Packages, error) {
-	panic("not implemented yet")
+func (i *Indexer) watchIndices(ctx context.Context) {
+	logger := util.Logger()
+	logger.Debug("Watch indices for changes")
+
+	var err error
+	t := time.NewTicker(i.options.WatchInterval)
+	defer t.Stop()
+	for {
+		logger.Debug("watchIndices: start")
+
+		err = i.updateIndex(ctx)
+		if err != nil {
+			logger.Error("can't update index file", zap.Error(err))
+		}
+
+		logger.Debug("watchIndices: finished")
+		select {
+		case <-ctx.Done():
+			logger.Debug("watchIndices: quit")
+			return
+		case <-t.C:
+		}
+	}
+}
+
+func (i *Indexer) updateIndex(ctx context.Context) error {
+	logger := util.Logger()
+	logger.Debug("Update indices")
+
+	bucketName, rootStoragePath, err := extractBucketNameFromURL(i.options.PackageStorageBucketInternal)
+	if err != nil {
+		return errors.Wrapf(err, "can't extract bucket name from URL (url: %s)", i.options.PackageStorageBucketInternal)
+	}
+
+	storageCursor, err := loadCursor(ctx, i.storageClient, bucketName, rootStoragePath)
+	if err != nil {
+		return errors.Wrap(err, "can't load latest cursor")
+	}
+
+	if storageCursor.Current == i.cursor {
+		logger.Info("cursor is up-to-date", zap.String("cursor.current", i.cursor))
+		return nil
+	}
+	logger.Info("cursor will be updated", zap.String("cursor.current", i.cursor), zap.String("cursor.next", storageCursor.Current))
+
+	anIndex, err := loadSearchIndexAll(ctx, i.storageClient, bucketName, rootStoragePath, *storageCursor)
+	if err != nil {
+		return errors.Wrapf(err, "can't load the search-index-all index content")
+	}
+	logger.Info("Downloaded new search-index-all index", zap.String("index.packages.size", fmt.Sprintf("%d", len(anIndex.Packages))))
+
+	refreshedList, err := transformSearchIndexAllToPackages(*anIndex)
```

Comment:
> Storage Indexer uses a goroutine to regularly hit the cloud bucket for cursor changes. When the cursor changes, the indexer pulls a new index from the bucket and transforms into `packages.Packages`.

Oh, after https://github.com/elastic/package-registry/pull/813#discussion_r876109194 I understood that we were going to use some kind of [watch API](https://cloud.google.com/storage/docs/object-change-notification) :slightly_smiling_face: If we are going to be polling the cursor, I would think again about using authenticated HTTP instead of the storage-specific client.

> I wanted to skip writing yet another model class and depend on `json.RawMessage`. I will try to apply the same approach to the Package Storage Infra. With manifests present in the manifest, the indexer will create a `ProxyPackageFileSystem` to use the `packages.NewPackage` to build the `Package` structure (a hackish method, but no need to rewrite the model).

Umm, the [current model we have for packages](https://github.com/elastic/package-registry/blob/2779e4089251076414b39c4ef43725cd1e9be752/packages/package.go#L55) is quite similar to the [model in the index](https://github.com/elastic/package-registry/pull/814/files#diff-5806dad16fb8aa7d4d9851ce935f31909f7af99e3d3070acbd111c59cf175f6dL18), have you tried to unmarshal directly to them? Maybe it works. I think I would prefer that to faking a file system over the index json.
Even if it doesn't work as is now, we could still evaluate the changes needed to adapt the current model, or the index, so the index works with current model in the registry.

If the model in the index cannot be aligned with the model here in the registry, I think I would still prefer to keep both models, unmarshal to the index one, and then convert it to the one in the registry. It may be cumbersome at first but easier to understand and maintain in the future.

> I'm not sure what to do about streaming other files. I plan for the `ProxyPackageFileSystem` to contain a streaming logic so that it can pull stuff from the public bucket. So far, the hardest part is to wire the spaghetti through `packages.NewPackage` and `ServeFile`.

For files in the storage, ideally the registry could answer with a redirect to the resource in GCP, so it doesn't serve files directly. We would need to check though if the http client in current versions of Kibana would follow these redirects.
We may need to add something in the current model to differentiate local from "external" packages, "external" packages wouldn't have a fs builder, but would have some way of translating the file name to urls, and do the redirect.

If http client in Kibana follows redirects, I would try to follow this approach: try to use the current model, but adapt it for "external" packages to serve files with redirects instead of from fake filesystems. This change to serve files with redirects could be done in a separated PR.

Reply:
Let me share an update on this:
1. I used marshalers to load the index fetched from the storage bucket.
2. I will focus on streaming package data in a follow-up. Otherwise, the size of this PR will grow up.
3. Re watch API:

> Should you use Object change notification?
Generally, you should not use Object change notification. The recommended tool for generating notifications that track changes to objects in your Cloud Storage buckets is [Pub/Sub notifications for Cloud Storage](https://cloud.google.com/storage/docs/pubsub-notifications), because it's faster, more flexible, easier to set up, and more cost-effective. For a step-by-step guide to setting up Pub/Sub notifications for Cloud Storage, see [Configuring Pub/Sub notifications for Cloud Storage](https://cloud.google.com/storage/docs/reporting-changes).

I understand that it's recommended to use a Pub/Sub mechanism, but not sure if we want to use another cloud service (stronger coupling with GCP). Do you think that we should rather use a basic HTTP client with Oauth? I admit that using the GCP Go client is really convenient and we may introduce potential bugs if we go with signed HTTP requests. Is it an attempt to make it universal/storage independent? 

Let me know what you think.

Reply:
> I admit that using the GCP Go client is really convenient and we may introduce potential bugs if we go with signed HTTP requests. Is it an attempt to make it universal/storage independent?

Using more generic HTTP would allow us to use more generic HTTP tooling and libraries, reducing dependencies and possibly simplifying development and testing. We could still use the library to initiate the HTTP client, keep its tokens updated and so on, but the operations could be generic.

A consequence of this would be effectively that this would be more storage independent, yes. I wouldn't consider this a priority, or a motivation for this, but this would be a nice to have :slightly_smiling_face:

But I agree that if this complicates things, and the GCP library doesn't expose an authenticated client we can reuse, the effort is probably not worth it.

Regarding pub/sub, we can leave this for a future optimization, if anytime needed.

Reply:
Thanks for ideas!